rad2/src/commands/profiles.rs

```diff
@@ -0,0 +1,127 @@
+use std::fs::{remove_file, File};
+use std::io;
+
+use failure::Fail;
+use serde_yaml as yaml;
+use structopt::StructOpt;
+
+use librad2::meta::profile::UserProfile;
+use librad2::paths::Paths;
+
+use crate::editor;
+
+#[derive(StructOpt)]
+/// Manage user profiles
+pub enum Commands {
+    /// Create a new profile
+    New { name: String },
+    /// Edit an existing profile
+    Edit { name: String },
+    /// Show an existing profile
+    Show { name: String },
+    /// Delete an existing profile
+    Delete { name: String },
+    /// List all profiles
+    List,
+}
+
+#[derive(Debug, Fail)]
+pub enum Error {
+    #[fail(display = "Profile `{}` already exists", 0)]
+    AlreadyExists(String),
+
+    #[fail(display = "Profile `{}` does not exist", 0)]
+    DoesNotExist(String),
+
+    #[fail(display = "{}", 0)]
+    Editor(editor::Error),
+
+    #[fail(display = "{}", 0)]
+    Yaml(yaml::Error),
+
+    #[fail(display = "{}", 0)]
+    Io(io::Error),
+}
+
+impl From<editor::Error> for Error {
+    fn from(err: editor::Error) -> Self {
+        Self::Editor(err)
+    }
+}
+
+impl From<yaml::Error> for Error {
+    fn from(err: yaml::Error) -> Self {
+        Self::Yaml(err)
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(err: io::Error) -> Self {
+        Self::Io(err)
+    }
+}
+
+pub fn run(paths: Paths, cmd: Commands, _verbose: bool) -> Result<(), Error> {
```

Comment:
How do you mean? It’s just a bunch of enums

Reply:
I meant that you make it a method:
```rust
impl Commands {
  pub fn run(self, paths: Paths, _verbose: bool) -> Result<(), Error> {}
}
```

Sorry I shouldn't have said `for` up above.

Reply:
Ya klar, but I would still need to pattern match on `self`, so it’s more of a syntactical thing. Maybe it’s nostalgia, but I kind of try to use plain functions wherever possible =]

Reply:
I, for one, accept our new Rust overlords.

I get what you mean, but I've found that I actually confuse myself when I try to consume a value with what I expected to be a method, but then I actually wrote a function. 

Another benefit is that if you want to change it later to pass by reference, then method calls don't need to be updated but function calls would need to be updated with `&`.

Some small things, so this is more a suggestion and seeing why choose one over the other :)

Reply:
I... am not convinced. I’m using this `structopt` thing, which magically turns these enum fields into subcommands. It then assembles the whole structure at once, and what’s left is to destructure it and run some effects at the leaves.

What you are saying would make sense if we’d stick the additional data that is passed to the function (and which comes from the outer options structure) inside the local commands - but that would mean I need two types, one for options parsing, and the other to actually run it.

We could stick “global” options like `verbose` into their own struct perhaps, so we only have to pass around one parameter.

I don’t understand the comment about references: are you saying that if I have

```rust
fn f(foo: Foo)
```

and I want to pass `foo` by reference, I need to change `f`s signature, whereas with

```rust
impl Whatever {
    fn f(foo: Foo)
}
```

the caller gets to choose?

Reply:
Ah I see what you mean and this way probably makes more sense to achieve that.

My first point was that I would  have something like:
```rust
struct Foo {}
fn consume(bar: Bar, foo: Foo) {}
```

And later on I would write `foo.consume(bar)` and then realise I wrote a function, but it's more common to have methods on structs in Rust. So my brain was broken.

My other point is that if you have:
```rust
struct Foo {}
impl Foo {
  pub fn consume(self, bar: Bar) {}
}

// cosnuming foo
foo.consume(bar)
```

if you then changed it to `consume(&self, bar: Bar)` you don't have to change `foo.consume(bar)`. But you'd have to add the `&` if you had `fn consume(bar: Bar: foo: &Foo)`. Not that that's a big deal or anything, just one less compiler error and a little more paint on the bikeshed :upside_down_face:

Reply:
>  I wrote a function, but it's more common to have methods on structs in Rust

Hm, I can name quite a few places in the standard library where things are implemented as a top-level function. Where convenient, they’re aliased as methods on one or more of the arguments. I guess it’s a feature that you don’t have to pick exactly one receiver?

It’s funny that you say “consume” - I always find myself having to think about move semantics. When a value is _consumed_, it doesn’t reappear where it went in, at best at the other end. Where I can’t seem to decide is when I need an owned copy - do I leave the `clone` to the caller, or do it where it’s needed? I guess the answer is to allocate as late as possible. Or something.

Reply:
> Hm, I can name quite a few places in the standard library where things are implemented as a top-level function. Where convenient, they’re aliased as methods on one or more of the arguments. I guess it’s a feature that you don’t have to pick exactly one receiver?

Hmm, fair enough. It might just be that I saw a few too many anecdotes on my side.

> It’s funny that you say “consume” - I always find myself having to think about move semantics

Ya, I kind of think of it in terms of: "Is this thing allowed to exist again outside of this function?" but then ya, you have `clone` as well. Mostly I just err on the side of using references as much as possible because they're cheaper to pass around and then you can re-use the values later. But that's mostly just rule of thumb and then it all depends on the context